singlem/appraiser.py

```diff
@@ -366,6 +366,147 @@ def mean(l):
 
+    def streaming_appraise(self, **kwargs):
+        '''Streaming appraisal function that sinks results to files immediately.'''
+        genome_otu_table_collection = kwargs.pop('genome_otu_table_collection')
+        metagenome_otu_table_collection = kwargs.pop('metagenome_otu_table_collection')
+        sequence_identity = kwargs.pop('sequence_identity', None)
+        output_found_in = kwargs.pop('output_found_in', False)
+        window_size = kwargs.pop('window_size')
+        binned_otu_table_io = kwargs.pop('binned_otu_table_io', None)
+        unbinned_otu_table_io = kwargs.pop('unbinned_otu_table_io', None)
+        threads=kwargs.pop('threads', 1)
+
+        if len(kwargs) > 0:
+            raise Exception("Unexpected arguments detected: %s" % kwargs)
+
+        logging.info("Using streaming mode for appraisal")
+        binned_otu_table_header = True
+        unbinned_otu_table_header = True
+        for otu_table, found in self._appraise_streaming(
+            metagenome_otu_table_collection,
+            genome_otu_table_collection,
+            sequence_identity,
+            output_found_in,
+            window_size,
+            threads=threads
+        ):
+            if output_found_in:
+                if found:
+                    if binned_otu_table_io:
+                        otu_table.write_to(binned_otu_table_io, fields_to_print=otu_table.fields, print_header=binned_otu_table_header)
+                        binned_otu_table_header = False
+                else:
+                    if unbinned_otu_table_io:
+                        otu_table.write_to(unbinned_otu_table_io, fields_to_print=otu_table.fields, print_header=unbinned_otu_table_header)
+                        unbinned_otu_table_header = False
+            else:
+                if found:
+                    if binned_otu_table_io:
+                        otu_table.write_to(binned_otu_table_io, print_header=binned_otu_table_header)
+                        binned_otu_table_header = False
+                else:
+                    if unbinned_otu_table_io:
+                        otu_table.write_to(unbinned_otu_table_io, print_header=unbinned_otu_table_header)
+                        unbinned_otu_table_header = False
+
+            if binned_otu_table_io:
+                binned_otu_table_io.flush()
+            if unbinned_otu_table_io:
+                unbinned_otu_table_io.flush()
+
+    def _appraise_streaming(self, metagenome_otu_table_collection,
+                                found_otu_collection,
+                                sequence_identity,
+                                output_found_in,
+                                window_size,
+                                threads=1):
+        '''Streaming version of appraise to handle large OTU tables with minimal memory usage.'''
+
+        def chunk_collection(iterator, chunk_size):
```

Comment:
there's a version of this in bird_tool_utils, make sense to use that instead?
https://github.com/wwood/bird_tool_utils-python/blob/main/bird_tool_utils/__init__.py#L97

Reply:
Why does it fill the last chunk with None?